src/expr/impl/src/scalar/case.rs

```diff
@@ -106,6 +107,152 @@ impl Expression for CaseExpression {
     }
 }
 
+/// With large scale of simple form match arms in case-when expression,
+/// we could optimize the `CaseExpression` to `ConstantLookupExpression`,
+/// which could significantly facilitate the evaluation of case-when.
+#[derive(Debug)]
+struct ConstantLookupExpression {
+    return_type: DataType,
+    arms: HashMap<ScalarImpl, BoxedExpression>,
+    fallback: Option<BoxedExpression>,
+    const_expression: Option<BoxedExpression>,
+}
+
+impl ConstantLookupExpression {
+    fn new(
+        return_type: DataType,
+        arms: HashMap<ScalarImpl, BoxedExpression>,
+        fallback: Option<BoxedExpression>,
+        const_expression: Option<BoxedExpression>,
+    ) -> Self {
+        Self {
+            return_type,
+            arms,
+            fallback,
+            const_expression,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Expression for ConstantLookupExpression {
+    fn return_type(&self) -> DataType {
+        self.return_type.clone()
+    }
+
+    async fn eval(&self, input: &DataChunk) -> Result<ArrayRef> {
+        let input_len = input.capacity();
+        let mut builder = self.return_type().create_array_builder(input_len);
+
+        // TODO: Ensure if the last column matches the actual result generated from the operand
+        // Get the last column
+        let column_len = input.columns().len();
+        let column = input.column_at(column_len - 1);
```

Comment:
Why do we get the last column instead of the first column? IIUC, the first column is the operand column.

Reply:
The final evaluated result would appear at the end of the input `DataChunk`, e.g.,
```sql
select case c1 + c2 * c3 + c4 when (...) then (...) else (...) end from t1;
```

Reply:
In my mind, we should use the `operand` expression to evaluate the input chunk to get the operand value, which should be used to lookup the hash table.